src/ode/man/man1/mklinks.1
.\"
.\" @OSF_FREE_COPYRIGHT@
.\" COPYRIGHT NOTICE
.\" Copyright (c) 1992, 1991, 1990  
.\" Open Software Foundation, Inc. 
.\"  
.\" Permission is hereby granted to use, copy, modify and freely distribute 
.\" the software in this file and its documentation for any purpose without 
.\" fee, provided that the above copyright notice appears in all copies and 
.\" that both the copyright notice and this permission notice appear in 
.\" supporting documentation.  Further, provided that the name of Open 
.\" Software Foundation, Inc. ("OSF") not be used in advertising or 
.\" publicity pertaining to distribution of the software without prior 
.\" written permission from OSF.  OSF makes no representations about the 
.\" suitability of this software for any purpose.  It is provided "as is" 
.\" without express or implied warranty. 
.\"
.\"
.\" HISTORY
.\" $Log: mklinks.1,v $
.\" Revision 1.8.6.1  1993/09/14  00:41:29  marty
.\" 	CR # 623
.\" 	[1993/09/14  00:39:31  marty]
.\"
.\" Revision 1.8.4.4  1993/04/21  18:42:16  damon
.\" 	CR 421. Updated for 2.3
.\" 	[1993/04/21  18:39:38  damon]
.\" 
.\" Revision 1.8.4.3  1993/03/31  16:47:25  marty
.\" 	Minor edits
.\" 	[1993/03/31  16:46:45  marty]
.\" 
.\" Revision 1.8.4.2  1993/03/31  16:45:34  marty
.\" 	OT # 263 , updated man page
.\" 	[1993/03/31  16:45:14  marty]
.\" 
.\" Revision 1.8.2.2  1992/12/03  17:24:17  damon
.\" 	ODE 2.2 CR 183. Added CMU notice
.\" 	[1992/12/03  17:10:16  damon]
.\" 
.\" Revision 1.8  1991/12/05  21:15:25  devrcs
.\" 	Changes from ODE 2.1 Man page review.
.\" 	[1991/11/03  18:39:12  damon]
.\" 
.\" 	Upgraded to Tools II interface
.\" 	[91/01/07  09:35:36  randyb]
.\" 
.\" 	Documented the -over option for copy.
.\" 	[90/11/07  13:57:20  randyb]
.\" 
.\" Revision 1.6  90/12/06  14:21:15  devrcs
.\" 	Cleanup copyright and history log comments.
.\" 	[90/11/15  11:46:30  gm]
.\" 
.\" Revision 1.5  90/10/07  21:56:50  devrcs
.\" 	Pre-OSF1 changes
.\" 
.\" $EndLog$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH mklinks 1 4/21/93
.ds ]W ODE 2.3
.SH NAME
mklinks \- populate the sandbox with links to files in a backing tree.
.SH SYNOPSIS
.IP \fBmklinks\fP 9
[\fB-copy\fP [\fB-over\fR]] [\fB-norecurse\fP] [\fB-refresh\fP]
[\fIODE opts\fR] \fIdirectory\fR | [\fIabs opts\fR]
.IP "\fIabs opts\fR:" 9
\fB-link_from \fIsource-directory\fP \fB-link_to \fInew-directory\fP
\fB-cmp\fR \fB-query\fR
.IP "\fIODE opts\fR:" 10
[\fB-info -auto -noauto -quiet -normal -verbose -usage -version\fP]
.br
[\fB-sb \fIsandbox\fR] [\fB-rc \fIrcfile\fP]
.SH FLAGS
.IP "\fB-copy\fP" 
Make actual copies rather than links.
The modification times of the files that are copied are preserved
for the benefit of make.
.IP "\fB-over\fP" 
Force copy to overwrite files which already exists.
WARNING: this option can overwrite files the user has checked-out.
.IP "\fB-norecurse\fP" 
Links files in the target directory only; does not recursively populate
subdirectories.
.IP "\fB-refresh\fP" 
Existing links will be replaced with new links.
.P
.ti 2
\fIabs opts\fR:
.IP "\fB-link_from \fP\fIsource-directory\fP"
Use \fIsource-directory\fP as the path of the directory
to make links from.
This is for general functionality of the tool and is not normally
used in sandboxes.
.IP "\fB-link_to \fP\fInew-directory\fP"
Use \fInew-directory\fP as the path to the directory to contain the
new links.
This is for general functionality of the tool and is not normally
used in sandboxes.
.IP "\fB-cmp\fP"
Compare existing files against the corresponding files in the backing tree and
replace identical files with links.
.IP "\fB-query\fP" 
Used with the -cmp flag. Queries the user for each file before it is deleted
and replaced with a link.
.P
.ti 2
\fIODE opts\fR:
.IP "\fB\-info\fR"
\fB-info\fR gives the user information on what \fBmklinks\fR would do
without actually doing anything.
The \fB-info\fR option works in conjunction with the verbosity options.
.IP "\fB\-auto -noauto\fR"
\fB-auto\fR eliminates any prompting \fBmklinks\fR would do.
If set, \fBmklinks\fR will take the default action without asking.
\fB-noauto\fR toggles \fB-auto\fR off and is most often used to turn
off \fB-auto\fR when it is set in the user's rc file.
.IP "\fB\-quiet\fR"
One of three levels of verbosity.
\fB-quiet\fR prints only the information which the user must always
receive including all fatal error messages.
.IP "\fB\-normal\fR"
One of three levels of verbosity.
\fB-normal\fR is the default verbosity and prints out the significant
steps the program is taking and warnings.
It also prints out all information \fB-quiet\fR prints.
.IP "\fB\-verbose\fR"
One of three levels of verbosity.
\fB-verbose\fR prints out detailed information on what steps the program
is taking and all diagnostic errors.
It also prints out all information \fB-quiet\fR and \fB-normal\fR prints.
.IP "\fB\-usage\fR"
Prints a brief usage message.
.IP "\fB\-version\fR"
Prints the revision information.
.IP "\fB\-sb \fIsandbox\fR"
By default, \fBmklinks\fR uses the current sandbox as determined by the
environment variable \fBSANDBOX\fR or, if that is not set, by
the default sandbox in the user's rc file.
If this option is given, the specified sandbox will be used.
.IP "\fB-rc \fIrcfile\fR"
By default, \fBmklinks\fR uses \fB$HOME/.sandboxrc\fR to determine the
current sandbox.
If this option is given, the specified user's rc file will be used.
.P
\fBmklinks\fR provides a quick and easy way to link or copy directories
recursively.
By default it knows about the sandbox structure and puts links in place
using that knowledge.
.SH DESCRIPTION
.P
The \fBmklinks\fP command creates symbolic
links in the users sandbox to source files in the backing
chain.
.P
\fBmklinks\fR identifies the files to link in one of two ways;
files relative to the current working directory of a sandbox and
files relative to the base source directory of a sandbox.
If the file name given to \fBmklinks\fR starts with a slash, ("/"),
the file to link to is assumed to be located relative to the 
source base directory.
If the file name does not begin with a slash,
then the link is created relative to the current working directory.
.P
If the users specifies a directory, then \fBmklinks\fR will attempt
to create links for all files in the directory as well as files in 
subdirectories.
.P
The file to link to (in the backing tree) must already exist.
If, however, the file does not exist in the sandbox,
it will be created.
.P
Options which apply to creating links in a sandbox are:
-copy, -over, and -norecurse.
.P
If, instead of automatically linking a sandbox with its backing tree,
the user wishes to link two arbitrary directories, he can do so by using
the \fB-link_from\fR and \fB-link_to\fR options together.
These two options are used as a pair and the \fBfile\fR entry
given with sandbox linking is omitted.
\fBmklinks\fR will return an error if only one of these options is given.
.P
Each option is followed by the absolute or relative path to the
appropriate directory.
If the directory path is absolute (starts with a slash)
it is taken from the root.
If it does not start with a slash, it is assumed to begin at the current
working directory.
.P
These two options can be used in conjunction with all the option
previously mentioned plus -cmp and -query.
.P
In a sandbox that already contains some links, \fBmklinks\fP will add any
new links required.
With the -cmp and -query options, \fBmklinks\fP can check or update sandboxes
that already contain some links.
.P
.SH EXAMPLES
.P
For purposes of these examples, the sandbox is assumed to be
\fB/sandbox/suzieq/song\fR.
.in +5
.sp
mklinks /src/usr/bin/yacc
.P
Links all files and directories, recursively, from
/sandbox/suzieq/song/link/src/usr/bin/yacc to
/sandbox/suzieq/song/src/usr/bin/yacc.
It does not matter where the current working directory is when this
command is called.
.P
The same result is achieved with these commands:
.in +5
.sp
.nf
cd /sandbox/suzieq/song/src/usr/bin/yacc
mklinks .
.fi
.in
.P
When a relative path is given, \fBmklinks\fP expects the user to be in
the sandbox.
.P
If the links are to be created in other than the current sandbox,
use the \fI-sb\fP flag:
.in +5
.sp
mklinks -sb /sandbox/suzieq/memories /src/usr/bin/yacc
.in
.PP
The \fI-norecurse\fP option stops \fBmklinks\fP from linking sub-directories.
To link the objects in kernel/bin without descending the sub-directories,
the command would be:
.in +5
.sp
mklinks -norecurse /obj/pmax/kernel/bin
.in
.P
To make local copies of the tools, use:
.in +5
.sp
mklinks -copy /tools
.in
.P
To link two arbitrary directories, for example the src in someone's sandbox
to the user's home directory, use the -link_ options.
.in +5
.sp
mklinks -link_from /sandbox/suzieq/song/src -link_to ~/src
.in
.P
These two options must always be used together.
.SH FILES
.PP
.IP \fB$HOME/.sandboxrc\fP 
is used to obtain the path to the sandbox.
.PP
.SH RELATED INFORMATION
.PP
sandboxrc(5),
mksb(1),
resb(1).



src/ode/lib/libode/mklinks.c
/*
 * @OSF_FREE_COPYRIGHT@
 * COPYRIGHT NOTICE
 * Copyright (c) 1992, 1991, 1990  
 * Open Software Foundation, Inc. 
 *  
 * Permission is hereby granted to use, copy, modify and freely distribute 
 * the software in this file and its documentation for any purpose without 
 * fee, provided that the above copyright notice appears in all copies and 
 * that both the copyright notice and this permission notice appear in 
 * supporting documentation.  Further, provided that the name of Open 
 * Software Foundation, Inc. ("OSF") not be used in advertising or 
 * publicity pertaining to distribution of the software without prior 
 * written permission from OSF.  OSF makes no representations about the 
 * suitability of this software for any purpose.  It is provided "as is" 
 * without express or implied warranty. 
 */
/*
 * HISTORY
 * $Log: mklinks.c,v $
 * Revision 1.8.4.3  1993/04/08  19:53:13  damon
 * 	CR 446. Clean up include files
 * 	[1993/04/08  19:53:01  damon]
 *
 * Revision 1.8.4.2  1993/01/13  16:43:49  damon
 * 	CR 382. Removed INC_DIRENT NO_DIRENT S_ISDIR S_ISLNK S_ISREG
 * 	[1993/01/05  21:02:21  damon]
 * 
 * Revision 1.8.2.5  1992/12/03  17:21:16  damon
 * 	ODE 2.2 CR 183. Added CMU notice
 * 	[1992/12/03  17:08:26  damon]
 * 
 * Revision 1.8.2.4  1992/11/06  16:23:09  damon
 * 	CR 296. Removed TRUE/FALSE defs
 * 	[1992/11/06  16:22:59  damon]
 * 
 * Revision 1.8.2.3  1992/09/24  19:01:55  gm
 * 	CR282: Made more portable to non-BSD systems.
 * 	[1992/09/21  20:35:03  gm]
 * 
 * Revision 1.8.2.2  1991/12/31  17:01:26  damon
 * 	ported to sparc_sunos
 * 	[1991/12/31  16:48:01  damon]
 * 
 * Revision 1.7.1.2  1991/12/17  15:26:47  damon
 * 	Ported to hp300_hpux
 * 	[1991/12/17  14:31:42  damon]
 * 
 * Revision 1.7  1991/12/05  21:05:22  devrcs
 * 	Removed include of dirent.h
 * 	[91/02/03  11:50:00  damon]
 * 
 * 	Added # include <dirent.h> explicitly.
 * 	[91/02/03  11:33:23  damon]
 * 
 * 	Changed sdm to ode; std_defs.h to  odedefs.h
 * 	[91/01/10  11:46:18  randyb]
 * 
 * 	Correct copyright; clean up lint input; added ui_print function.
 * 	Also a fairly thorough once-over to make this routine comply
 * 	with the changes in mklinks.c the program.
 * 	[91/01/08  12:16:33  randyb]
 * 
 * 	Add test so -copy would overwrite existing files even if they exist if
 * 	the -over option is thrown.
 * 
 * 	Added -debug level to the output so -v and -debug do not give the same
 * 	information.
 * 	[90/11/07  16:53:23  randyb]
 * 
 * Revision 1.5  90/10/07  20:03:41  devrcs
 * 	Added EndLog Marker.
 * 	[90/09/28  20:09:36  gm]
 * 
 * Revision 1.4  90/08/09  14:23:20  devrcs
 * 	Moved here from usr/local/sdm/lib/libsb.
 * 	[90/08/05  12:32:20  gm]
 * 
 * Revision 1.3  90/07/17  12:36:50  devrcs
 * 	More changes for gcc.
 * 	[90/07/04  22:30:53  gm]
 * 
 * Revision 1.2  90/05/24  23:12:31  devrcs
 * 	Added missing set of braces.
 * 	[90/05/10  19:54:39  gm]
 * 
 * 	Created.
 * 	[90/05/06            gm]
 * 
 * $EndLog$
 */

#include <dirent.h>
#include <sys/errno.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>

#include <ode/interface.h>
#include <ode/odedefs.h>
#define  CERROR	(-1)
    extern int 	  errno;
    extern char * prog;

int
makelink ( stp, isnew, isdir, overwrite, namebuf, srcpath,
	   curpath, srcpref, cmpfunc, linkfunc )
    
    struct 	stat *stp;
    int 	  isnew, isdir, overwrite;
    char 	* namebuf, *srcpath, *curpath, *srcpref;
    int        (* cmpfunc)();
    int        (* linkfunc)();

{
    char 	  buf1 [MAXPATHLEN];
    struct 	stat st, bst;
    int 	  status;

  ui_print ( VDEBUG, "makelink ( stp %x isnew %d isdir %d overwrite %d\n",
      		      stp, isnew, isdir, overwrite );
  ui_print ( VCONT,
		 "namebuf %s\n   srcpath %s\n   curpath %s\n   srcpref %s )\n",
	          namebuf, srcpath, curpath, srcpref ? srcpref : NULL );

  if (isdir) {
    if (!isnew) {
      if (stat(curpath, &st) < 0) {
	if (errno != ENOENT) {
	  ui_print ( VFATAL, "stat %s: %s\n", curpath, strerror(errno));
	  return ( CERROR );
	}
      }
      else if (S_ISDIR(st.st_mode))
        return(FALSE);
    }

    if (cmpfunc != NULL)
      return(TRUE);

    if (mkdir(curpath, 0777) < 0) {
      ui_print ( VFATAL, "mkdir %s: %s\n", curpath, strerror(errno));
	return(CERROR);
    }

    ui_print ( VDETAIL, "%s: created directory\n", namebuf);
    return(TRUE);
  }

  if (( !isnew && lstat ( curpath, &st ) == 0 ) && !overwrite ) {
    if (cmpfunc == NULL) {
      ui_print ( VDETAIL, "%s already exists\n", namebuf );
      return(FALSE);
    } /* if */

    ui_print ( VDEBUG, "%s: mode %o\n", curpath, st.st_mode );

    if (S_ISLNK(st.st_mode)) {
      if (stat(curpath, &st) != 0) {
	ui_print ( VFATAL, "stat %s: %s\n", curpath, strerror(errno));
	return(CERROR);
      }
    }

    ui_print ( VCONT, "%s: mode %o\n", curpath, st.st_mode );

    if (!S_ISREG(st.st_mode))
      return(FALSE);

    if (stp == NULL) {
      stp = &bst;
      if (lstat(namebuf, stp) != 0) {
	ui_print ( VFATAL, "lstat %s: %s\n", namebuf, strerror(errno));
	return(CERROR);
      }
    }
    
    ui_print ( VDEBUG, "%s: mode %o\n", namebuf, stp->st_mode );

    if (S_ISLNK(stp->st_mode)) {
      if (stat(namebuf, stp) != 0) {
	ui_print ( VFATAL, "stat %s: %s\n", namebuf, strerror(errno));
	return(CERROR);
      }
    }
    
    ui_print ( VCONT, "%s: mode %o\n", namebuf, stp->st_mode );

    if (!S_ISREG(stp->st_mode))
      return(FALSE);
    
    ui_print ( VCONT, "%s: size %d vs %d\n",
			curpath, st.st_size, stp->st_size );

    if (st.st_size != stp->st_size || st.st_size == 0)
      return(FALSE);
    status = (*cmpfunc)(namebuf, curpath);
    ui_print ( VCONT, "%s: cmp %d\n", namebuf, status );

    if (status != 0)
      return((status == CERROR) ? CERROR : FALSE);

    (void) unlink(curpath);
  } /* if */
  
  else {
    if (( !isnew && errno != ENOENT ) && !overwrite ) {
      ui_print ( VFATAL, "lstat %s: %s\n", curpath, strerror(errno));
      return(CERROR);
    }
    if (cmpfunc != NULL)
      return(FALSE);
  }

  if (srcpref != NULL && *srcpref != '\0') {
    if (concat(buf1, sizeof(buf1), srcpref, srcpath, NULL) == NULL) {
      ui_print ( VFATAL, "%s: path too long\n", buf1);
      return(FALSE);
    }

    ui_print ( VDEBUG, "linkfunc(%s,\n   %s)\n", buf1, curpath );

    if ((*linkfunc)(buf1, curpath) == CERROR)
      return(CERROR);
  } /* if */
  
  else {
    ui_print ( VDEBUG, "linkfunc(%s,\n   %s)\n", srcpath, curpath );

    if ((*linkfunc)(srcpath, curpath) == CERROR)
      return(CERROR);
  }

  ui_print ( VDETAIL, "%s: created\n", namebuf);
  return(FALSE);
}  								 /* makelink */



void
descend ( ndirs, isnew, query, recurse, overwrite, namebuf, endname, srcpath,
	  endsrc, curpath, endcur, srcpref, endpref, cmpfunc, linkfunc )

    int 	  ndirs;
    int 	  isnew;
    int 	  query;
    int 	  recurse;
    int 	  overwrite;
    char 	* namebuf,
		* endname;
    char 	* srcpath, 
		* endsrc;
    char 	* curpath, 
		* endcur;
    char 	* srcpref, 
		* endpref;
    int        (* cmpfunc)();
    int        (* linkfunc)();

{
    struct 	stat st;
    /* struct 	direct *dp; 			   old name, OSF uses dirent */
    struct 	dirent *dp;
    DIR 	* dirp;
    int 	  subisnew;
    char 	* np, *sp, *cp, *pp;
    char 	  buf [MAXPATHLEN];

  ui_print ( VDEBUG, "namebuf: %s\n", namebuf );
  ui_print ( VCONT,
	    "descend ( ndirs %d isnew %d query %d recurse %d overwrite %d\n",
	     ndirs, isnew, query, recurse, overwrite );
  ui_print ( VCONT, "srcpath %s\n   curpath %s\n   srcpref %s)\n",
		     srcpath, curpath, srcpref ? srcpref : NULL);

  if ((dirp = opendir(namebuf)) == NULL) {
    perror(namebuf);
    return;
  }

  for (dp = readdir(dirp); dp != NULL; dp = readdir(dirp)) {
    if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0) {
      continue;
    }
    np = concat(endname, namebuf+MAXPATHLEN-endname, "/", dp->d_name, NULL);

    if (np == NULL) {
      ui_print ( VWARN, "%s: path too long\n", namebuf);
      continue;
    }
    sp = concat(endsrc, srcpath+MAXPATHLEN-endsrc, "/", dp->d_name, NULL);

    if (sp == NULL) {
      ui_print ( VWARN, "%s: path too long\n", srcpath);
      continue;
    }
    cp = concat(endcur, curpath+MAXPATHLEN-endcur, "/", dp->d_name, NULL);

    if (cp == NULL) {
      ui_print ( VWARN, "%s: path too long\n", curpath);
      continue;
    }

    if (ndirs == 0) {
      (void) makelink (( struct stat *)NULL, isnew, FALSE, overwrite, namebuf,
			 srcpath, curpath, srcpref, cmpfunc, linkfunc );
      continue;
    }

    if (lstat(namebuf, &st) != 0) {
      perror(namebuf);
      continue;
    }

    if (!S_ISDIR(st.st_mode)) {
      (void) makelink ( &st, isnew, FALSE, overwrite, namebuf, srcpath,
			curpath, srcpref, cmpfunc, linkfunc );
      continue;
    }

    ndirs--;

    if (query) {
      (void) concat (buf, sizeof(buf), "Link directory ", namebuf, " ?", NULL);
      if (!getbool(buf, TRUE))
	continue;
    }

    if (!recurse)
      continue;

    subisnew = makelink( &st, isnew, TRUE, overwrite, namebuf, srcpath,
			 curpath, srcpref, cmpfunc, linkfunc);

    if (subisnew == CERROR)
      continue;

    if (subisnew && cmpfunc != NULL)
      continue;

    if (srcpref != NULL) {
      pp = concat (endpref, srcpref+MAXPATHLEN-endpref, "../", NULL);

      if (pp == NULL) {
	ui_print ( VFATAL, "%s: path too long\n", srcpref);
	return;
      }
    }
    
    else
      pp = NULL;

    descend ( st.st_nlink-2, subisnew, query, recurse, overwrite, namebuf, np,
	      srcpath, sp, curpath, cp, srcpref, pp, cmpfunc, linkfunc );

    if (endpref != NULL)
      *endpref = '\0';
  } /* for */

  (void) closedir(dirp);
}  								  /* descend */



mklinks ( isnew, query, recurse, overwrite, srcpath, curpath, srcpref,
	  cmpfunc, linkfunc)

    int 	  isnew, query, recurse, overwrite;
    char 	* srcpath, *curpath, *srcpref;
    int        (* cmpfunc) ();
    int        (* linkfunc) ();

{
    char 	  namebuf[MAXPATHLEN];
    struct 	stat st;

  namebuf[0] = '.';
  namebuf[1] = '\0';

  if ( stat ( namebuf, &st ) < 0 )
    uquit( ERROR, FALSE, "stat %s: %s\n", srcpath, strerror(errno));

  descend ( st.st_nlink - 2, isnew, query, recurse, overwrite,
	    namebuf, namebuf + 1,
	    srcpath, srcpath + strlen(srcpath),
	    curpath, curpath + strlen(curpath),
	    (srcpref != NULL) ? srcpref : NULL,
	    (srcpref != NULL) ? srcpref + strlen(srcpref): NULL,
	    cmpfunc, linkfunc );
}  								  /* mklinks */
